### ISSUE TITLE
Use 127.0.0.1 instead of localhost in tests

### ISSUE THREAD

**Issue by Kriechi** [mitmproxy/mitmproxy#1971]
Use `127.0.0.1` in all tests instead of `localhost` (which might not be available on all systems).

ref #1966

**Comment by mhils** [comment]
`127.0.0.1` won't work for IPv6, so it's a bit more tricky than that. Still we should fix that in general.

**Comment by synthesizerpatel** [comment]
I hate to be 'that guy', but RFC-6761 calls out localhost as being special and so much depends on it being there.. I can't stop myself.

> 6.3.  Domain Name Reservation Considerations for "localhost."
> 
>    The domain "localhost." and any names falling within ".localhost."
>    are special in the following ways:
> 
>    1.  Users are free to use localhost names as they would any other
>        domain names.  Users may assume that IPv4 and IPv6 address
>        queries for localhost names will always resolve to the respective
>        IP loopback address.

tl;dr If /etc/hosts doesn't at least have '127.0.0.1 localhost' the fabric of space time will rupture.

**Comment by mhils** [comment]
@synthesizerpatel, I do agree with you, but what kind of UX do you propose for systems where this isn't the case? See #1966.

**Comment by synthesizerpatel** [comment]
Every operating system has a localhost entry after a fresh install. Windows, Linux, Android, Plan9, etc.

As a goof I went and looked at the Python-3.4.4 source to try and drive the point home.

`find . -name "*.py" -exec grep localhost {} \; | wc -l`
`286`

If it's good enough for Guido. ;)


For the sake of argument - and since you asked for a solution.

There are a couple dumb / klunky ways you could work around it:

* Run the tests in a Docker container. Most folks do these days because it's easy (albeit lazy).

* If you wanted to be super ninja you could monkey-patch the builtin socket module so that gethostbyname() and its ilk could catch the failure - try to figure out which iface you want to connect to if they have both ipv4/ipv6 configured.. I don't know how well that would work. But it might. It'd be a horrible solution though.

**Comment by mosajjal** [comment]
if we're trying to reach `localhost`, it shouldn't matter if we're using ipv4 or ipv6. We can always use `::1` as our test address, right?

**Comment by mhils** [comment]
> We can always use `::1` as our test address, right?

I think we often still bind to IPv4 only, so no. Thinking longer about this, @synthesizerpatel has a strong point with RFC6761 and it's probably the best thing for now to just keep things as-is.
